### PR TITLE
fix: eagerly listen to `TileLayer.reset` stream to ensure handler is fired

### LINKED ISSUE #1808: [BUG] `TileLayer.reset` internal listener never triggers
### What is the bug?

I was trying to use reset to trigger reload one of my TileLayer, but I found that the subsciption of reset never be initialized.
In lib/src/layer/tile_layer.dart, the subscription of reset dosen't be triggered properly:

```Dart
class _TileLayerState extends State<TileLayer> with TickerProviderStateMixin {
  bool _initializedFromMapCamera = false;

  final _tileImageManager = TileImageManager();
  late TileBounds _tileBounds;
  late var _tileRangeCalculator =
      TileRangeCalculator(tileSize: widget.tileSize);
  late TileScaleCalculator _tileScaleCalculator;

  // We have to hold on to the mapController hashCode to determine whether we
  // need to reinitialize the listeners. didChangeDependencies is called on
  // every map movement and if we unsubscribe and resubscribe every time we
  // miss events.
  int? _mapControllerHashCode;

  StreamSubscription<TileUpdateEvent>? _tileUpdateSubscription;
  Timer? _pruneLater;

  late final _resetSub = widget.reset?.listen((_) {
    _tileImageManager.removeAll(widget.evictErrorTileStrategy);
    _loadAndPruneInVisibleBounds(MapCamera.of(context));
  });

  // This is called on every map movement so we should avoid expensive logic
  // where possible.
  @override
  void didChangeDependencies() {
    super.didChangeDependencies();

    final camera = MapCamera.of(context);
    final mapController = MapController.of(context);
.......
```

Because keyword "late" was used for _resetSub, and it never be accessed which means it will not be initialized. I tried to add one line "print(_resetSub);", then this subscription works. I think that might be bug.

### How can we reproduce it?

1. Include a `TileLayer` in your Flutter application.
2. Pass a proper `reset` stream for the `TileLayer`, exactly same logic with the flutter map demo app.
3. Trigger a reset event to observe the issue. (add break point inside the reset subsription will find the program doesn't get into it)

### Do you have a potential solution?

Access it somewhere might be able to solve this problem.

### Platforms

macOS, web

### Severity

Minimum: Allows normal functioning

### PR DESCRIPTION
**Pull request by JaffaKetchup**
Also added refreshing of reset listener when reset stream changes within widget lifecycle.

Fixes #1808. However, the behaviour is still not correct, because the tile loading logic is incorrect. This looks like #1813:
 * If the `urlTemplate` does not change, tiles are evicted but not loaded until the map is moved (and possibly the tiles are not evicted correctly because they appear to reload quickly, so maybe from cache, but `reset` should empty cache)
 * If the `urlTemplate` does change, tiles are evicted but then instantly reloaded (see above), and new tiles load in gradually replacing/overlaying the old ones - the old tiles should not become visible in-between

Therefore, this will need a follow up PR.

### CODE REVIEW

File: lib/src/layer/tile_layer/tile_layer.dart [hunk 1]
```diff
@@ -345,10 +344,14 @@ class _TileLayerState extends State<TileLayer> with TickerProviderStateMixin {
   StreamSubscription<TileUpdateEvent>? _tileUpdateSubscription;
   Timer? _pruneLater;
 
-  late final _resetSub = widget.reset?.listen((_) {
-    _tileImageManager.removeAll(widget.evictErrorTileStrategy);
-    if (mounted) _loadAndPruneInVisibleBounds(MapCamera.of(context));
-  });
+  StreamSubscription<void>? _resetSub;
```
Comment (josxha) [inline lib/src/layer/tile_layer/tile_layer.dart:347]: Stream needs to get canceled in dispose().

File: lib/src/layer/tile_layer/tile_layer.dart [hunk 2]
```diff
@@ -706,6 +714,11 @@ class _TileLayerState extends State<TileLayer> with TickerProviderStateMixin {
 
   bool _outsideZoomLimits(num zoom) =>
       zoom < widget.minZoom || zoom > widget.maxZoom;
+
+  void _resetStreamHandler(void _) {
+    _tileImageManager.removeAll(widget.evictErrorTileStrategy);
+    if (mounted) _loadAndPruneInVisibleBounds(MapCamera.of(context));
```
Comment (josxha) [inline lib/src/layer/tile_layer/tile_layer.dart:720]: Does this handle the tile buffer around the visible tiles, too?

**Review by josxha — COMMENTED** [review]
Nice pull request @JaffaKetchup!👍Added two small comments.
Could be nice to have a test for this, but no pressure. 😄

File: lib/src/layer/tile_layer/tile_layer.dart (hunk 1 above)
Comment (JaffaKetchup) [reply lib/src/layer/tile_layer/tile_layer.dart:347]: Ah, forgot to do this, but did it in update :D

File: lib/src/layer/tile_layer/tile_layer.dart (hunk 2 above)
Comment (JaffaKetchup) [reply lib/src/layer/tile_layer/tile_layer.dart:721]: This doesn't change the reset logic, it only ensures the handler is fired. Another PR is needed to look into this and fix #1813.

File: lib/src/layer/tile_layer/tile_layer.dart (hunk 1 above)
Comment (JaffaKetchup) [reply lib/src/layer/tile_layer/tile_layer.dart:347]: Stream is already cancelled in `dispose`:  https://github.com/fleaflet/flutter_map/blob/8136ef0fc0b75b759ea5e02eeaa2e1eac3dca245/lib/src/layer/tile_layer/tile_layer.dart#L469

**Comment by JaffaKetchup** [comment]
We can add a more detailed test for the pruning etc. logic when #1813 is fixed.

File: lib/src/layer/tile_layer/tile_layer.dart (hunk 1 above)
Comment (josxha) [reply lib/src/layer/tile_layer/tile_layer.dart:347]: My bad sorry, haven't seen it while reviewing.

**Review by josxha — APPROVED** [review]
👍

**Pull request merged**